cmd/kubeadm/app/phases/addons/dns/dns.go

```diff
@@ -244,3 +275,108 @@ func createDNSService(dnsService *v1.Service, serviceBytes []byte, client client
+// in the form of Proxy for the CoreDNS Corefile.
+func translateStubDomainOfKubeDNSToProxyCoreDNS(dataField string, kubeDNSConfigMap *v1.ConfigMap) (string, error) {
+	if proxy, ok := kubeDNSConfigMap.Data[dataField]; ok {
+		stubDomainData := make(map[string][]string)
+		var proxyStanzaList string
+
+		err := json.Unmarshal([]byte(proxy), &stubDomainData)
+		if err != nil {
+			return "", fmt.Errorf("failed to parse JSON from 'kube-dns ConfigMap: %v", err)
+		}
+		var proxyStanza []interface{}
+
+		for domain, proxyIP := range stubDomainData {
+			strings.Join(proxyIP, " ")
+			pStanza := map[string]interface{}{}
+			pStanza["keys"] = []string{domain + ":53"}
+			pStanza["body"] = [][]string{
+				{"errors"},
+				{"cache", "30"},
+				{"proxy", ".", strings.Join(proxyIP, " ")},
+			}
+
+			proxyStanza = append(proxyStanza, pStanza)
+		}
+		stanzasBytes, err := json.Marshal(proxyStanza)
+		if err != nil {
+			return "", err
+		}
+
+		outputJSON, err := caddyfile.FromJSON(stanzasBytes)
+		if err != nil {
+			return "", err
+		}
+		// This is required to format the Corefile, otherwise errors due to bad yaml format.
+		output := strings.NewReplacer("\n\t", "\n        ", "\"", "", "\n}", "\n    }", "\n\n", "\n    ").Replace(string(outputJSON))
+		proxyStanzaList = coreDNSStanzaFormat + output + coreDNSStanzaFormat
+
+		return proxyStanzaList, nil
+	}
+	return "", nil
+}
+
+// translateUpstreamNameServerOfKubeDNSToUpstreamProxyCoreDNS translates UpstreamNameServer Data in kube-dns ConfigMap
+// in the form of Proxy for the CoreDNS Corefile.
+func translateUpstreamNameServerOfKubeDNSToUpstreamProxyCoreDNS(dataField string, kubeDNSConfigMap *v1.ConfigMap) (string, error) {
+	if upstreamValues, ok := kubeDNSConfigMap.Data[dataField]; ok {
+		var upstreamProxyIP []string
+
+		err := json.Unmarshal([]byte(upstreamValues), &upstreamProxyIP)
+		if err != nil {
+			return "", fmt.Errorf("failed to parse JSON from 'kube-dns ConfigMap: %v", err)
+		}
+
+		coreDNSProxyStanzaList := strings.Join(upstreamProxyIP, " ")
+		return coreDNSProxyStanzaList, nil
+	}
+	return "/etc/resolv.conf", nil
+}
+
+// translateFederationsofKubeDNSToCoreDNS translates Federations Data in kube-dns ConfigMap
+// to Federation for CoreDNS Corefile.
+func translateFederationsofKubeDNSToCoreDNS(dataField, coreDNSDomain string, kubeDNSConfigMap *v1.ConfigMap) (string, error) {
+	if federation, ok := kubeDNSConfigMap.Data[dataField]; ok {
+		var (
+			federationStanza []interface{}
+			fData            []string
+		)
+		federationData := make(map[string]string)
+
+		err := json.Unmarshal([]byte(federation), &federationData)
+		if err != nil {
+			return "", fmt.Errorf("failed to parse JSON from kube-dns ConfigMap: %v", err)
+		}
+
+		for name, domain := range federationData {
+			fData = append(fData, name+" "+domain)
+		}
+		fStanza := map[string]interface{}{}
+		fStanza["body"] = [][]string{
+			{strings.Join(fData, "\n       ")},
```

Comment:
Relative indentation assumptions is a terrible thing to do, why?

Reply:
The indentation isn’t strictly necessary for the conversion, but I insisted on it because after the upgrade is done and the user wants to modify the configmap, it ensures that the configmap is in a readable format and this makes it easier for the user to interact.

Reply:
I am not entirely sure this is an essential kind of optimisation. Please add a comment to clarify the intention.